gax-grpc/src/main/java/com/google/api/gax/grpc/SafeShutdownManagedChannel.java

```diff
@@ -0,0 +1,150 @@
+ * shutting down.
+ *
+ * <p>This class is not thread-safe. Caller must synchronize in order to ensure no new calls if safe
+ * shutdown has started.
+ *
+ * <p>Package-private for internal use.
+ */
+class SafeShutdownManagedChannel extends ManagedChannel {
+  private ManagedChannel delegate;
+  private AtomicInteger outstandingCalls = new AtomicInteger(0);
+  private volatile boolean isShutdownSafely = false;
+
+  SafeShutdownManagedChannel(ManagedChannel managedChannel) {
+    this.delegate = managedChannel;
+  }
+
+  /**
+   * Safely shutdown channel by checking that there are no more outstanding calls. If there are
+   * outstanding calls, the last call will invoke this method again when it complete
+   */
+  void shutdownSafely() {
+    isShutdownSafely = true;
+    if (outstandingCalls.get() == 0) {
+      delegate.shutdown();
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdown() {
+    delegate.shutdown();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isShutdown() {
+    return delegate.isShutdown();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isTerminated() {
+    return delegate.isTerminated();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdownNow() {
+    delegate.shutdownNow();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return delegate.awaitTermination(timeout, unit);
+  }
+
+  /**
+   * Wrap client call to decrement outstandingCalls and shutdown channel if necessary when it
+   * completes
+   *
+   * @param call call to be wrapped
+   * @return the wrapped call
+   */
+  private <ReqT, RespT> ClientCall<ReqT, RespT> clientCallWrapper(ClientCall<ReqT, RespT> call) {
+    return new SimpleForwardingClientCall<ReqT, RespT>(call) {
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        Listener<RespT> forwardingResponseListener =
+            new SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onClose(Status status, Metadata trailers) {
+                // decrement in finally block in case onClose throws an exception
+                try {
+                  super.onClose(status, trailers);
+                } finally {
+                  if (outstandingCalls.decrementAndGet() == 0 && isShutdownSafely) {
+                    shutdownSafely();
+                  }
+                }
+              }
+            };
+        super.start(forwardingResponseListener, headers);
+      }
+    };
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+      MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+    // increment after client call in case newCall throws an exception
+    assert (!isShutdownSafely);
```

Comment:
asserts are not very useful in java, they are disabled by default. Generally people use guava's preconditions for the same effect: `Precondition.checkState(!isShutdownSafely)`

Reply:
Also, I know you mentioned it in the class javadoc, but its worth calling it out here as well. Please add a comment, explaining that this code expects to be called in a synchronized block to avoid race conditions